### PR TITLE
Calculate function and validate functions under Cycle.Uuid

### LINKED ISSUE #1684: Validations/calculations/metacache by cycle

### PR DESCRIPTION
**Pull request by sorja**
Resolves #1684 

- [x] row.calculateFn [926578a](https://github.com/openforis/fra-platform/pull/1687/commits/926578a2d4577619f483e39683bb51784c08748d)
- [x] row.validateFns [926578a](https://github.com/openforis/fra-platform/pull/1687/commits/926578a2d4577619f483e39683bb51784c08748d)
- [x] metacache [5f9d3bc](https://github.com/openforis/fra-platform/pull/1687/commits/5f9d3bcebdda05c46e477eb623d34ea1f97151f2)
- [x] col.calculateFn [aaa3fcd](https://github.com/openforis/fra-platform/pull/1687/commits/aaa3fcd890b64b4366079fecaf41f18243ef7381)

row.{calculateFn,validateFns}
Screenshot 2022-11-08 at 15 24 52

assessment.metaCache
image

col.props.calculateFn
Screenshot 2022-11-08 at 16 54 50

### CODE REVIEW

File: src/server/controller/cycleData/persistNodeValue/calculateDependantNodes/calculateDependantNodes.ts [hunk 1]
```diff
@@ -41,18 +41,18 @@ export const calculateDependantNodes = async (props: Props, client: BaseProtocol
...
         user,
       }
 
-      if (row.props.calculateFn) {
+      if (row.props.calculateFn?.[cycle.uuid]) {
         // make sure in target table there's a matching column
         if (row.cols.find((c) => c.props.colName === colName)) {
           // eslint-disable-next-line no-await-in-loop
-          const node = await calculateNode({ ...evaluateProps, expression: row.props.calculateFn }, client)
+          const node = await calculateNode({ ...evaluateProps, expression: row.props.calculateFn[cycle.uuid] }, client)
```
Comment (minotogna) [inline src/server/controller/cycleData/persistNodeValue/calculateDependantNodes/calculateDependantNodes.ts:55]: also col.props.calculateFn should be by cycle

**Review by minotogna — CHANGES_REQUESTED** [review]
really good pr 👍 
just col calculateFn prop should by by cycle

File: src/server/controller/cycleData/persistNodeValue/calculateDependantNodes/calculateDependantNodes.ts (hunk 1 above)
Comment (sorja) [reply src/server/controller/cycleData/persistNodeValue/calculateDependantNodes/calculateDependantNodes.ts:55]: good catch, wasn't 100% sure about it, will do

File: src/meta/assessment/row.ts [hunk 2]
```diff
@@ -45,7 +47,8 @@ export interface RowProps {
   readonly?: boolean
   type: RowType
   variableName?: string
-  validateFns?: Array<string>
+  //            Record<cycleUuid, Array<string>>
+  validateFns?: Record<string, Array<string>>
```
Comment (minotogna) [inline src/meta/assessment/row.ts]: remove comment and use CycleUuid

File: src/test/sectionSpec/colSpec.ts [hunk 3]
```diff
@@ -1,6 +1,7 @@
 import BigNumber from 'bignumber.js'
 
 import { ColStyle, Label } from '@meta/assessment'
+import { CycleUuid } from '@meta/assessment/cycle'
```
Comment (minotogna) [inline src/test/sectionSpec/colSpec.ts]: CycleUuid must be imported from '@meta/assessment'

File: src/test/sectionSpec/rowSpec.ts [hunk 4]
```diff
@@ -1,4 +1,5 @@
 import { VariableCache } from '@meta/assessment'
+import { CycleUuid } from '@meta/assessment/cycle'
```
Comment (minotogna) [inline src/test/sectionSpec/rowSpec.ts]: CycleUuid must be imported from '@meta/assessment'

**Review by minotogna — CHANGES_REQUESTED** [review]

**Review by minotogna — APPROVED** [review]

**Review by rfontanarosa — APPROVED** [review]

**Pull request merged**